ADR/0040-worspace-design-proposal.md

```diff
@@ -0,0 +1,172 @@
+
+Given these limitations in Kubernetes, it becomes clear that we can benefit with a mechanism for abstracting workspaces from their underlying namespaces. The approach proposed here to best represent the workspace concept is to extend the kubernetes API via a new CRD. The Konflux user will be able to interact with workspaces as a resource of kubernetes. They will be able to execute all CRUD operations to manage their workspace lifecycle.
+
+### Architecture
+
+The workspace team has factored these limitations in while developing features on Konflux downstream by creating a layer between the Konflux User Interfaces and the KubeSaw components that provide namespace as a service (NSaaS) layer. We understand that there are multiple ways to implement a namespace provisioning service depending on the constraints of the system. Our first implementation is KubeSaw, but as we redefined some constraints on Konflux deployments, other backends for a workspace became viable. See the diagram below for the different backends that can be implemented at the moment.
+
+#### Diagram
+
+![](../diagrams/ADR-0040/konflux-workspaces.png)
+
+#### Description
+
+The diagram depicts 3 backend implementations for the Workspace Controller: KubeSaw, Local and GitOps. All backends end up managing namespaces, their initial resources required, and connectivity between these namespaces to authenticated identities.
+
+##### KubeSaw
+
+This is the current implementation downstream running today in our KubeSaw deployments. It leverages the KubeSaw host operator, member operator and registration service components. It is designed to scale with multiple member cluster support. This implementation will only be possible with OpenShift clusters, since there is no vanilla support of K8S at the moment. Implementing support for non-OpenShift Kubernetes deployments is scheduled for the future.
+
+##### Local
+
+This implementation assumes Konflux runs in single cluster mode. It is a straightforward implementation of local namespace provisioning, bootstrapping and user allocation. It will require a workspace config object that describes the default resources to set during the bootstrapping phase.
+
+##### GitOps
+
+This implementation provisions the namespace in a target cluster, but the bootstrapping is delegated to a GitOps provider like ArgoCD. A new application will be spawned on the git repo and ArgoCD will track those resources and synchronize them with the cluster. This is the current approach using upstream at the moment.
+
+### Features by backend
+
+The table below expands with more details on the features that each backend can provide and which features will not be available.
+
+| Feature | KubeSaw | Local | GitOps |
+| :---- | :---- | :---- | :---- |
+| Single-cluster | ✅ | ✅ | ⚠️ (ArgoCD) |
+| Multi-cluster | ✅ | ❌ | ✅ |
```

Comment:
what does it mean multicluster ? multicluster namespace provisioning? how traffic will be routed to the different clusters?

Reply:
A user's workspace would be assigned to a single cluster, and the set of all workspaces for a konflux installation could be distributed across multiple distinct clusters.  We're not looking to have a workspace living on multiple clusters.

* * *

With the kubesaw backend, we would solve the traffic routing problem with registration-service.

Reply:
so traffic routing isn't something related to the workspaces abstraction? if so why it should be mentioned in this adr?

Reply:
While kubesaw would implement traffic routing for us, the GitOps backend might require a different solution.  I'm not sure if we could get registration-service working standalone in this use case, but if not, this routing problem is something that the workspaces abstraction is going to need to deal with.

Either way, it's something that workspaces will need to care about, even if it does punt a solution to another provider.

Reply:
Important thing to understand is that the multi-cluster functionality is important for the user & namespace management. It shouldn't have any impact on the operators or APIs provisioned in the clusters. 

Let's take an example of DevSandbox:
* there are 3 member clusters installed with a long list of operators. None of the operators is multi-cluster aware - they work within the scope of the cluster only
* plus there is one more member cluster (provisioned with some specific configuration) that has only OpenShift AI operator installed and is used only for a specific use-case.

Each installation of the operator in each member cluster doesn't know about each other. And when you you are provisioned in any of the member clusters, then you have no clue about the others and what API is available in the other member clusters.

Having all data located in one cluster makes the admin-work much easier as there is only one source of truth. 

And the proxy (and registration-service) makes the usage of the multi-cluster environment easier for the end-users (or 3rd party services) - there is only one endpoint they need to talk to and there is only one URL the users need to get into to be redirected to the correct member-cluster or to know which clusters it has access to. Or you can configure your internal UI so it talks to the proxy instead of each of the member clusters.

In similar way, you may have one or more clusters installed & configured with Konflux - the number of Konflux installations is equal to the number of member clusters. Konflux doesn't have to be multi-cluster or proxy aware, but the whole environment (user & namespace management) can still benefit from the multi-cluster setup.